### PR TITLE
[Android]Decouple handler for process js runtime error in native code

### PR DESCRIPTION
**Pull request by sospartan**
* Decouple  `JavascriptExceptionHandler` from `ExceptionsManagerModule`.
*  Allow customize error handler in production.  It will be helpful for developers to adding a ideal process when JS error happened (caused by runtime JS error ,e.g. fatal server response,not JS evaluation) in native code. They can :  
  * Let users to report these errors by email or some error-report service.
  * Restart react-native app by receate JS context.

cc @mkonicek

### CODE REVIEW

**Comment by sospartan** [comment]
@foghina I see your commit 7871abf907fcf24c4167be8a95854082cc5df5db  which deal with `JSExecutionException` .  This PR is similar except dealing with JavascriptException. Are you mind to help review this one?

**Comment by sospartan** [comment]
I've updated this PR to resolve the conflict.
@foghina We are almost fixing the thing with same code at nearly same time. That is cool!

**Comment by foghina** [comment]
@sospartan this looks good, except you can already do what you want with `setNativeModuleCallExceptionHandler` that, as you pointed out, I introduced in 7871abf. Basically, what happens when an uncaught JS exception is thrown is:

* `ExceptionsManager.js` catches it and passes it to the native ExceptionsManager module.
* If in debug mode, the native module shows a redbox
* If in release mode, the native module rethrows the exception as a `JavascriptException`. As this is thrown during a standard `@ReactMethod` call, this is caught ([here](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/queue/MessageQueueThreadHandler.java#L33)) and passed to the registered `NativeModuleCallExceptionHandler` ([here](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/bridge/CatalystInstanceImpl.java#L438)).

Basically, you can put all your release mode exception handling logic in the `NativeModuleCallExceptionHandler`. This will actually give you access to handling even more exceptions than just `JavascriptException` - basically anything that happens during a native method call. If you *don't* want to handle other exceptions, you can just rethrow them and crash the app.

Or maybe I'm just missing the point. Do you have a use case where using a `NativeModuleCallExceptionHandler` doesn't work?

**Comment by sospartan** [comment]
@foghina I see. The name 'NativeModuleCallExceptionHandler' is misleadingly. I use to think this handler only catch the 'exceptions throw in native code & JS evaluating error',but the JS runtime error.  
I'll close this PR, and I suggest to change the name since it catch everything.
